### PR TITLE
Avoid restarting eventing-controller if NATS cluster was restarted

### PR DESCRIPTION
**Pull request by radufa**
**Description**

Changes proposed in this pull request:

- The controller is  reacting now for K8S events  related to the lifecycle of NATS cluster instance
- The Kyma subscriptions are injected now at restarting of NATS cluster
- The eventing controller should not be  manually restarted if NATS cluster was restarted

**Related issue(s)**

See: https://github.com/kyma-project/kyma/issues/11376

### CODE REVIEW

File: components/eventing-controller/pkg/handlers/nats.go [hunk 1]
```diff
@@ -117,29 +112,19 @@ func (n *Nats) SyncSubscription(sub *eventingv1alpha1.Subscription, cleaner even
...
 	}
 	return false, nil
 }
 
 func (n *Nats) DeleteSubscription(subscription *eventingv1alpha1.Subscription) error {
-	subNsName := types.NamespacedName{
-		Namespace: subscription.Namespace,
-		Name:      subscription.Name,
-	}
 	for k, v := range n.subscriptions {
```
Comment (themue) [inline components/eventing-controller/pkg/handlers/nats.go:126]: Even if `k` and `v` for key and value are technically correct I would appreciate `key` and `subscription`.

**Review by themue — APPROVED** [review]
I'm fine with the idea but would like to see some syntactical changes.

File: components/eventing-controller/pkg/handlers/nats.go [hunk 2]
```diff
@@ -159,6 +147,17 @@ func (n *Nats) DeleteSubscription(subscription *eventingv1alpha1.Subscription) e
 	return nil
 }
 
+func (n *Nats) GetInvalidSubscriptions() *[]types.NamespacedName {
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats.go:150]: Done

File: components/eventing-controller/pkg/handlers/nats.go [hunk 3]
```diff
@@ -197,3 +196,33 @@ func convertMsgToCE(msg *nats.Msg) (*cev2event.Event, error) {
 	}
 	return &event, nil
 }
+
+func getKeyPrefix(sub *eventingv1alpha1.Subscription) string {
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats.go:200]: changed to `createXyz()`

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 4]
```diff
@@ -172,3 +175,123 @@ func TestSubscription(t *testing.T) {
...
+func TestIsValidSubscription(t *testing.T) {
+	g := NewWithT(t)
+
+	natsPort := 5223
+	subscriberPort := 8080
+	subscriberReceiveURL := fmt.Sprintf("http://127.0.0.1:%d/store", subscriberPort)
+
+	// Start Nats server
+	natsServer := eventingtesting.RunNatsServerOnPort(natsPort)
+	//defer natsServer.Shutdown()
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats_test.go:188]: Removed

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 5]
```diff
@@ -172,3 +175,123 @@ func TestSubscription(t *testing.T) {
...
+
+	// Create Nats client
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats_test.go:190]: Done

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 6]
```diff
@@ -172,3 +175,123 @@ func TestSubscription(t *testing.T) {
...
+	g.Expect(len(*invalidNsn)).To(BeZero())
+
+	natsClient.connection.Close()
+
+	// the associated Nats subscription should not be valid anymore
+	err = checkIsNotValid(natsSub, t)
+	g.Expect(err).To(BeNil())
+
+	// shutdown NATS
+	natsServer.Shutdown()
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats_test.go:250]: Done

File: components/eventing-controller/pkg/handlers/nats.go (hunk 1 above)
Comment (radufa) [reply components/eventing-controller/pkg/handlers/nats.go:126]: Done

File: components/eventing-controller/reconciler/subscription-nats/reconciler.go [hunk 7]
```diff
@@ -243,3 +279,23 @@ func (r Reconciler) assertProtocolValidity(protocol string) error {
 	}
 	return nil
 }
+
+func (r Reconciler) syncInvalidSubscriptions(ctx context.Context) (ctrl.Result, error) {
+	handler, _ := r.Backend.(*handlers.Nats)
+	invalidSubs := handler.GetInvalidSubscriptions()
+	for _, v := range *invalidSubs {
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/reconciler/subscription-nats/reconciler.go:286]: changed

**Review by themue — APPROVED** [review]

**Pull request merged**